### PR TITLE
feat: unify agent run API with volume version override support

### LINKED ISSUE #255: refactor: unify agent run API endpoints into single entry point
## Summary

Refactor the current three separate API endpoints (`/api/agent/runs`, `/api/agent/runs/resume`, `/api/agent/runs/continue`) into a single unified endpoint that uses parameters to determine the run mode. This follows the original design intent from #240.

## Complete CLI Usage

The final CLI should support:

```bash
vm0 run agent-demo \
  --conversation conversationId \
  --volume-version theVolumeName=theVersion \
  --volume-version claude-system=latest \
  --volume-version vim=1a2b3c \
  --artifact-name theArtifactName \
  --artifact-version theArtifactVersion \
  "prompt"
```

### Key Features

1. **`--conversation`** - Resume from a specific conversation (session history)
2. **`--volume-version`** - Dynamically specify volume versions at run time (CRITICAL NEW FEATURE)
   - Can be specified multiple times for different volumes
   - Format: `volumeName=version`
   - `latest` gets the latest version
   - Specific version like `1a2b3c` pins to that version
3. **`--artifact-name`** - Artifact storage name
4. **`--artifact-version`** - Artifact version (default: latest)

### Shortcut Parameters

For convenience, `--checkpoint` and `--session` are "syntactic sugar" that expand to the above base parameters:

```bash
# These are equivalent:
vm0 run --checkpoint <checkpointId> "prompt"
vm0 run --conversation <conversationId> --volume-version <from-checkpoint> --artifact-version <from-checkpoint> "prompt"

# These are equivalent:
vm0 run --session <sessionId> "prompt"
vm0 run --conversation <conversationId> --artifact-version latest "prompt"
```

## Background

Issue #240 proposed a unified API design where all run modes share a single entry point. However, the current implementation created **three separate API endpoints**:
- `POST /api/agent/runs` - New run
- `POST /api/agent/runs/resume` - Resume from checkpoint
- `POST /api/agent/runs/continue` - Continue from session

### 1. Severe Code Duplication

The three route files share ~80% identical code:
- Authentication logic
- Run record creation
- Event sending (vm0_start, vm0_error)
- Async execution pattern
- Error handling

### 2. Scattered Logic

RunService has three separate context creation methods:
- `createRunContext()` - New run
- `createResumeContext()` - Checkpoint resume  
- `createContinueContext()` - Session continue

### 3. Inflexible

Adding new run modes requires creating new API endpoints.

### 4. Missing Dynamic Volume Version Support

Current implementation cannot dynamically adjust volume versions at run time - a critical feature for production use.

### Unified Request Interface

```typescript
interface CreateRunRequest {
  // High-level shortcuts (mutually exclusive)
  checkpointId?: string;        // Expand checkpoint snapshot parameters
  sessionId?: string;           // Expand session parameters (artifact version forced to "latest")
  
  // Low-level parameters (can be used directly or overridden)
  agentConfigId?: string;       // Agent config ID
  conversationId?: string;      // Conversation to resume from
  artifactName?: string;        // Artifact storage name
  artifactVersion?: string;     // Artifact version (default: "latest")
  dynamicVars?: Record<string, string>;  // Template variables
  volumeVersions?: Record<string, string>; // NEW: volume name -> version mapping
  
  // Required
  prompt: string;
}
```

### Parameter Resolution Flow

```
Request Parameters
    │
    ├─ Has checkpointId? → Expand checkpoint snapshot:
    │     - agentConfigId = checkpoint.agentConfigSnapshot.configId
    │     - conversationId = checkpoint.conversationId
    │     - artifactName = checkpoint.artifactSnapshot.artifactName
    │     - artifactVersion = checkpoint.artifactSnapshot.artifactVersion
    │     - dynamicVars = checkpoint.agentConfigSnapshot.templateVars
    │     - volumeVersions = checkpoint.volumeVersions (from snapshot)
    │
    ├─ Has sessionId? → Expand session data:
    │     - agentConfigId = session.agentConfigId
    │     - conversationId = session.conversationId
    │     - artifactName = session.artifactName
    │     - artifactVersion = "latest"  ← ALWAYS latest for session!
    │     - dynamicVars = session.templateVars
    │
    └─ Merge/override with explicit parameters (including volumeVersions!)
           │
           ↓
    Unified ExecutionContext
           │
           ↓
    e2bService.execute(context)
```

### Key Behavior Differences

| Run Mode | config | conversation | artifact version | volume versions |
|----------|--------|--------------|------------------|-----------------|
| New run | from request | none | from request (default: latest) | from request |
| `--checkpoint` | checkpoint snapshot | checkpoint snapshot | checkpoint snapshot | from snapshot (overridable) |
| `--session` | session's config | session's latest conversation | **always latest** | from request |

### Phase 1: Refactor RunService

1. Create unified `buildExecutionContext()` method:
```typescript
async buildExecutionContext(params: {
  checkpointId?: string;
  sessionId?: string;
  agentConfigId?: string;
  conversationId?: string;
  artifactName?: string;
  artifactVersion?: string;
  dynamicVars?: Record<string, string>;
  volumeVersions?: Record<string, string>;  // NEW
  prompt: string;
  runId: string;
  sandboxToken: string;
  userId: string;
}): Promise<ExecutionContext>
```

2. Update `ExecutionContext` type to include `volumeVersions`

3. Deprecate (but keep for now):
   - `createRunContext()`
   - `createResumeContext()`
   - `createContinueContext()`
   - `validateAgentSession()`

### Phase 2: Refactor API Endpoint

1. Update `POST /api/agent/runs` to accept unified request
2. Extract common logic into helper functions
3. Mark `/resume` and `/continue` as deprecated (keep working for backwards compatibility)

### Phase 3: Update CLI

1. Add `--volume-version` flag (repeatable) to `vm0 run`
2. Add `--conversation` flag to `vm0 run`
3. Update `vm0 run` to use unified API
4. Keep `vm0 run resume` and `vm0 run continue` as aliases

### Phase 4: Update E2B Service

1. Update e2b-service to handle `volumeVersions` parameter
2. Apply volume version overrides when mounting volumes

### Phase 5: Cleanup

1. Remove deprecated endpoints
2. Remove old RunService methods

## E2E Test Requirements

**Critical: All new CLI functionality must have E2E test coverage.**

### Tests to Add

```bash
# Test: Basic run with volume version
@test "VM0 run: specify single volume version" {
  run vm0 run agent-demo \
    --volume-version "my-volume=latest" \
    --artifact-name "$ARTIFACT_NAME" \
    "echo test"
  assert_success
}

# Test: Multiple volume versions
@test "VM0 run: specify multiple volume versions" {
  run vm0 run agent-demo \
    --volume-version "claude-system=latest" \
    --volume-version "vim=1a2b3c" \
    --artifact-name "$ARTIFACT_NAME" \
    "echo test"
  assert_success
}

# Test: Conversation resume with volume override
@test "VM0 run: resume conversation with volume version override" {
  # First run creates checkpoint
  run vm0 run agent-demo --artifact-name "$ARTIFACT_NAME" "echo initial"
  # Get conversation ID from checkpoint
  
  # Resume with different volume version
  run vm0 run agent-demo \
    --conversation "$CONVERSATION_ID" \
    --volume-version "my-volume=specific-version" \
    "echo continued"
  assert_success
}

# Test: Checkpoint with volume version override
@test "VM0 run: checkpoint resume with volume version override" {
  # Even when using checkpoint, can override volume versions
  run vm0 run --checkpoint "$CHECKPOINT_ID" \
    --volume-version "my-volume=different-version" \
    "echo test"
  assert_success
}

# Test: Session continue (should use latest artifact)
@test "VM0 run: session continue uses latest artifact" {
  run vm0 run --session "$SESSION_ID" "echo continued"
  assert_success
  # Verify artifact version is "latest"
}
```

### Existing Tests to Update

Update `e2e/tests/02-commands/t06-vm0-agent-session.bats` to cover:
- New `--conversation` flag
- New `--volume-version` flag
- Combined usage scenarios

### Core Changes
- `turbo/apps/web/src/lib/run/run-service.ts` - Add `buildExecutionContext()`
- `turbo/apps/web/src/lib/run/types.ts` - Add unified types, add `volumeVersions`
- `turbo/apps/web/app/api/agent/runs/route.ts` - Accept new parameters
- `turbo/apps/web/src/types/agent-run.ts` - Update request type
- `turbo/apps/web/src/lib/e2b/e2b-service.ts` - Handle `volumeVersions`

### CLI Changes
- `turbo/apps/cli/src/commands/run.ts` - Add `--volume-version`, `--conversation` flags
- `turbo/apps/cli/src/lib/api-client.ts` - Update `createRun()` method

### E2E Test Changes
- `e2e/tests/02-commands/t06-vm0-agent-session.bats` - Update existing tests
- Add new test file or test cases for volume version functionality

### Deprecate (Keep Working)
- `turbo/apps/web/app/api/agent/runs/resume/route.ts`
- `turbo/apps/web/app/api/agent/runs/continue/route.ts`

### Unit Tests
- [ ] `buildExecutionContext()` handles all parameter combinations
- [ ] `volumeVersions` correctly merges with defaults
- [ ] Checkpoint expansion includes volume versions
- [ ] Session expansion uses latest artifact

### Integration Tests
- [ ] API endpoint accepts new request format
- [ ] Volume versions passed to E2B service correctly

### E2E Tests
- [ ] `vm0 run agent --artifact-name name "prompt"` works (new run)
- [ ] `vm0 run agent --volume-version name=version "prompt"` works
- [ ] `vm0 run agent --volume-version a=v1 --volume-version b=v2 "prompt"` works
- [ ] `vm0 run agent --conversation <id> "prompt"` works
- [ ] `vm0 run --checkpoint <id> "prompt"` works (checkpoint resume)
- [ ] `vm0 run --checkpoint <id> --volume-version name=override "prompt"` works
- [ ] `vm0 run --session <id> "prompt"` works (session continue)
- [ ] `vm0 run resume <id> "prompt"` alias works
- [ ] `vm0 run continue <id> "prompt"` alias works
- [ ] Deprecated endpoints still work

## Related Issues

- #240 - Original design for Agent Session concept
- #243 - PR that implemented the current (non-unified) approach

## Labels

enhancement

### PR DESCRIPTION
**Pull request by lancy**
## Summary

- Consolidates three separate agent run API endpoints (`/runs`, `/runs/resume`, `/runs/continue`) into a single unified endpoint
- Adds `--volume-version` CLI flag for runtime volume version override
- Adds volume version snapshots to checkpoints for reproducibility

## CLI Usage

```bash
# New run with volume version override
vm0 run agent-demo \
  --artifact-name theArtifactName \
  --volume-version theVolumeName=theVersion \
  --volume-version claude-system=latest \
  "prompt"

# Resume from checkpoint with volume override
vm0 run resume <checkpointId> \
  --volume-version volumeName=version \
  "prompt"

# Continue session with volume override
vm0 run continue <sessionId> \
  --volume-version volumeName=version \
  "prompt"
```

## Test plan

- [ ] Run E2E tests for volume version override (`t07-vm0-volume-version-override.bats`)
- [ ] Verify `vm0 run resume` and `vm0 run continue` aliases still work
- [ ] Verify `--volume-version` flag correctly overrides volume versions
- [ ] Verify checkpoint stores volume versions for reproducibility

Closes #255

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: e2e/tests/02-commands/t07-vm0-volume-version-override.bats [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+}
+
+teardown() {
+    # Clean up temporary directory
+    if [ -n "$TEST_DIR" ] && [ -d "$TEST_DIR" ]; then
+        rm -rf "$TEST_DIR"
+    fi
+}
+
+@test "Build VM0 volume override test agent configuration" {
```
Comment (lancy) [inline e2e/tests/02-commands/t07-vm0-volume-version-override.bats]: Is this necessary? Why not define it in fixtures/configs?

File: turbo/apps/web/app/api/agent/runs/resume/route.ts [hunk 2]
```diff
@@ -1,188 +1,42 @@
...
+import { NextRequest, NextResponse } from "next/server";
 
 /**
  * POST /api/agent/runs/resume
- * Resume execution from a checkpoint
+ *
+ * @deprecated Use POST /api/agent/runs with { checkpointId, prompt } instead.
+ * This endpoint redirects to the unified runs API.
  */
 export async function POST(request: NextRequest) {
```
Comment (lancy) [inline turbo/apps/web/app/api/agent/runs/resume/route.ts]: I think there's no need to keep "redirect"; we can just delete it. Let's not do backward compatibility for now.

**Review by lancy — COMMENTED** [review]
I think we should add some more unit tests for the module's functions.

**Comment by lancy** [comment]
@lancy Thanks for the review feedback! I've added comprehensive unit tests as requested:

## Added Tests

### `run-service.test.ts` (20 tests)
- `calculateSessionHistoryPath` function tests (5 tests)
- `createRunContext` tests (2 tests)
- `buildExecutionContext` tests for all modes:
  - New run mode (3 tests)
  - Checkpoint resume mode (5 tests)
  - Session continue mode (5 tests)

### `storage-resolver.test.ts` (33 tests)
- `parseMountPath` function tests (9 tests)
- `replaceTemplateVars` function tests (8 tests)
- `resolveVolumes` function tests (16 tests)
  - Volume resolution with template variables
  - Volume version override behavior
  - Error handling for various edge cases

All 53 new tests pass. The tests focus on the key functions introduced or modified in this PR, particularly the unified `buildExecutionContext` function and volume version override support.
